evm_trace/geth.py

```diff
@@ -50,22 +49,78 @@ class TraceFrame(BaseModel):
     storage: Dict[HexBytes, HexBytes] = {}
     """Contract storage."""
 
+    contract_address: Optional[HexBytes] = None
+    """The address producing the frame."""
+
     @validator("pc", "gas", "gas_cost", "depth", pre=True)
     def validate_ints(cls, value):
         return int(value, 16) if isinstance(value, str) else value
 
-    @cached_property
+    @property
     def address(self) -> Optional[HexBytes]:
         """
         The address of this CALL frame.
         Only returns a value if this frame's opcode is a call-based opcode.
         """
 
-        if self.op not in CALL_OPCODES or CallType.CREATE.value in self.op:
-            # CREATE address is handled elsewhere when creating CallTreeNode.
-            return None
+        if not self.contract_address and (
+            self.op in CALL_OPCODES and CallType.CREATE.value not in self.op
+        ):
+            self.contract_address = HexBytes(self.stack[-2][-20:])
+
+        return self.contract_address
+
+
+def create_trace_frames(data: Iterator[Dict]) -> Iterator[TraceFrame]:
+    """
+    Get trace frames from ``debug_traceTransaction`` response items.
+    Sets the ``contract_address`` for CREATE and CREATE2 frames by
+    looking ahead and finding it.
+
+    Args:
+        data (Iterator[Dict]): An iterator of response struct logs.
+
+    Returns:
+        Iterator[:class:`~evm_trace.geth.TraceFrame`]
+    """
+
+    # NOTE: Use a new iter in case a list or something is passed in.
+    # This logic requires an iterator.
+    frames = iter(data)
+
+    for frame in frames:
+        if CallType.CREATE.value in frame.get("op", ""):
+            # Before yielding, find the address of the CREATE.
+            call_frames = []
+            start_depth = frame.get("depth", 0)
+            for next_frame in frames:
```

Comment:
Does it make sense to `itertools.tee` the frames at this point for the look ahead? Then drop it later and iterate on the original

Reply:
it doesn't make a difference but i can refactor to use itertools.tee if you like.

Reply:
i guess it does look simpler so ya that is good